File: homeworks/Nigmatullin/semesterProblems/AnDS/List.cpp
#include "List.h"
template<typename T> pElement<T>::pElement(pElement<T> *n, T el)
{
	data = el;
	next = n;
}

template<typename T> pElement<T>::pElement(int el)
{
	data = el;
	next = NULL;
}

template<typename T> pElement<T>::pElement()
{
	data = NULL;
	next = NULL;
}



template<typename T> List<T>::List()
{
	tail = NULL;
	head = tail;
	current = tail;
	size = 0;
}

template<typename T> void List<T>::append(T el)
{
	if (head == NULL)
	{
		head = new pElement<T>(el);
		tail = head;
	}
	else
	{
		tail->next = new pElement<T>(el);
		tail = tail->next;
	}
	size++;
}


template<typename T> void List<T>::prepend(T el)
{
	if (head == NULL)
	{
		pElement<T> *temp = new pElement<T>(head, el);
		head = temp;
		tail = head;
	}
	else
	{
		pElement<T> *temp = new pElement<T>(head, el);
		head = temp;
	}
	size++;
}

template<typename T> bool List<T>::del(int position)
{
	if (position > size - 1)
	{
		return false;
	}
	if (position == 0)
	{
		if (head != NULL)
		{
			pElement *temp = head;
			head = head->next;
			delete(temp);
			return true;
		}
		else
		{
			return false;
		}
	}
	else if (position == size - 1)
	{
		if (tail != NULL)
		{
			pElement *temp = head;
			for (int i = 0; i < position - 1; i++)
			{
				temp = temp->next;
			}
			delete(temp->next);
			temp->next = NULL;
			tail = temp;
			return true;
		}
		else
		{
			return false;
		}
	}
	else
	{
		pElement *temp = head;
		for (int i = 0; i < position - 1; i++)
		{
			temp = temp->next;
		}
		pElement *t = temp->next;

		temp->next = temp->next->next;
		delete(t);
		return true;
	}
}
template<typename T> bool List<T>::isEmpty()
{
	return (head == NULL);
}

template<typename T> int List<T>::Size()
{
	int i = 0;
	pElement<T> *temp = head;
	while (temp != NULL)
	{
		i++;
		temp = temp->next;
	}
	return i;
}

template<typename T> bool List<T>::insertElement(int position, T el)
{
	if (position > size)
	{
		return false;
	}
	if (position == 0)
	{
		if (head == NULL)
		{
			head = new pElement<T>(el);
			tail = tail;
			return true;
		}
		else
		{
			pElement<T> *temp = new pElement<T>(head, el);
			head = temp;
			return true;
		}
	}
	else if (position == size)
	{
		pElement<T> *temp = new pElement<T>(el);
		tail->next = temp;
		tail = tail->next;
		return true;
	}
	else
	{
		pElement<T> *temp = head;
		for (int i = 0; i < position - 1; i++)
		{
			temp = temp->next;
		}
		temp->next = new pElement<T>(temp->next, el);
		return true;
	}
}

File: homeworks/FedorovaOM/Semester_II/List/List.pas
Unit UList;

Interface

Const
  BolvankaValue = 0;

Type
  TData = Integer;
  PElement = ^TElement;
  TElement = Record
    data: TData;
    next: PElement;
  end;
  TList = Record
    head, tail, current: PElement;
    size: Integer;
  end;

function initElement(data: TData): PElement;
procedure initList(var list: TList);

procedure reset(var list: TList);
procedure next(var list: TList);
procedure moveTo(var list: TList; position: Integer);
procedure print(list: TList);

function isEmpty(list: TList): Boolean;
procedure append(var list: TList; data: TData);
procedure insertElement(var list: TList; position: Integer; el: PElement);
function get(var list: TList; position: Integer): TData;
procedure delete(var list: TList; position: Integer);

Implementation

function initElement(data: TData): PElement;
var
  el: PElement;
begin
  new(el);
  el^.data := data;
  el^.next := nil;
  Result := el;
end;

procedure initList(var list: TList);
begin
  list.head := initELement(BolvankaValue);
  list.tail := list.head;
  list.current := nil;
  list.size := 0;
end;

function isEmpty(list:TList): Boolean;
begin
  Result := list.size = 0;
end;

procedure append(var list: TList; data: TData);
begin
  list.tail^.next := initElement(data);
  list.tail := list.tail^.next;
  inc(list.size);
end;

procedure insertElement(var list: TList; position: Integer; el: PElement);
begin
  moveTo(list, position - 1);
  // el := initElement(data) // if we use data instead of el
  el^.next := list.current^.next;
  list.current^.next := el;
  inc(list.size);
end;

procedure print(list: TList);
begin
  reset(list);
  while list.current <> nil do begin
    write(list.current^.data,' ');
    next(list);
  end;
end;

procedure reset(var list: TList);
begin
  list.current := list.head^.next;
end;

procedure next(var list: TList);
begin
  list.current := list.current^.next;
end;

procedure moveTo(var list: TList; position: Integer);
var
  i: Integer;
begin
  reset(list);
  for i := 2 to position - 1 do
    next(list);
end;

function get(var list: TList; position: Integer): TData;
begin
  reset(list);
  moveTo(list, position);
  Result := list.current^.data;
end;

procedure delete(var list: TList; position: Integer);
var
  el: PElement;
begin
  reset(list);
  moveTo(list, position - 1);
  el := list.current^.next;
  list.current^.next := list.current^.next^.next; // el^.next
  dispose(el);
end;

end.
